Related product updates now being reflected in page previews

danielfmiranda opened the pull request:
# Description

<!-- Describe the PR here -->

Link to sample test page: https://foundation-s-11799-rela-3u9dyh.herokuapp.com/cms/pages/22/edit/ 
Test page credentials: admin2/admin2
Related PRs/issues: #11799


This PR introduces a dedicated template for previewing product pages, located at `previews/buyersguide/product_page.html`. 

It overrides the `{% block related_products %}` from the standard product page template to use a new template, `related_products_section.html`, which is filled with data retrieved through a new method named `preview_related_products`.

This new method leverages the existing utility `get_related_items` to fetch related product relationships directly from the CMS's in-memory state rather than the database. 

This addresses the issue where changes to related products within the CMS would not appear in the preview pane unless the page was published.

Lastly, this PR adds some tests for the new `preview_related_products` method.



# Steps to test

1. Visit https://foundation-s-11799-rela-3u9dyh.herokuapp.com/cms/pages/22/edit/ (credentials: admin2/admin2)
2. Scroll down to the related products section
3. Make a change
4. Note that the preview panel is updating accordingly
5. If everything is working as expected, testing is complete!

jhonatan-lopes commented on network-api/networkapi/wagtailpages/pagemodels/buyersguide/products.py:
```diff
@@ -1026,7 +1029,24 @@ def localized_related_products(self):
+            self.related_product_pages.all(),
+            "related_product",
+        )
+        # FIXME: This implementation does return the localized version of each product.
+        #        But, it is inefficient. It would be better to pull all products
+        #        for the correct locale at once. This would require the above returns
+        #        a queryset of the products (rather than a list) and that we have an
+        #        efficient way of pulling all items for a given locale.
+        #        See: https://github.com/MozillaFoundation/foundation.mozilla.org/issues/9509
```
We can change this comment for some more appropriate for this method to the tune of "we know this is inefficient but that's on purpose as we need to make sure that the preview renders the objects in memory"

danielfmiranda replied:
Done!

jhonatan-lopes commented on network-api/networkapi/templates/previews/product_page.html:
Can we move this to "previews/buyersguide/product_page.html" please?

danielfmiranda replied:
Thanks @jhonatan-lopes! I forgot to do that. The location has now been updated :+1:

jhonatan-lopes commented on network-api/networkapi/templates/fragments/buyersguide/related_products_section.html:
Awesome to have this in a separate fragment! :tada:

jhonatan-lopes commented on network-api/networkapi/templates/pages/buyersguide/product_page.html:
```diff
@@ -168,21 +168,14 @@ <h2 class="tw-h3-heading">{% trans "Tips to protect yourself" %}</h2>
-                    </section>
-                {% endif %}
-            {% endwith %}
+            {% block related_products %}
+                {% with related_products=product.localized_related_products %}
+                    {% if related_products %}
+                      {% include "fragments/buyersguide/related_products_section.html" with related_products=related_products %}
+                    {% endif %}
+                {% endwith %}
+            {% endblock related_products %}
```
Much more readable now :tada:

jhonatan-lopes reviewed (commented):
Hey @danielfmiranda, thanks for taking this work! It's all looking great! I've added some comments on where I think it can be improved

jhonatan-lopes commented on network-api/networkapi/wagtailpages/pagemodels/buyersguide/products.py:
```diff
@@ -1026,7 +1029,24 @@ def localized_related_products(self):
+        """
+        Fetches related product updates for CMS page previews.
+        """
+        related_products = orderables.get_related_items(
+            self.related_product_pages.all(),
+            "related_product",
+        )
```
I've just noticed something: this **does not** guarantee to order the related products by their `sort_order`

![image](https://github.com/MozillaFoundation/foundation.mozilla.org/assets/61277874/c0754639-80b4-4d39-8b77-39d349994bc7)

danielfmiranda replied:
Thanks for the callout @jhonatan-lopes!

[Here is a link ](https://github.com/MozillaFoundation/foundation.mozilla.org/pull/11844/files#r1483822553)to the solution I have implemented to make sure that they are always sorted by sort_order.

danielfmiranda commented on network-api/networkapi/utility/orderables.py:
```diff
@@ -1,15 +1,21 @@
 
 
-def get_related_items(queryset: "models.QuerySet", related_item_field: str) -> list["models.Model"]:
+def get_related_items(
+    queryset: "models.QuerySet", related_item_field: str, order_by: Optional[str] = None
+) -> list["models.Model"]:
     """
     Return list of the related items from the queryset.
 
     For each item in the queryset, the related item stored in `related_item_field` is
     returned.
 
     """
+    # Apply ordering if needed
+    if order_by:
+        queryset = queryset.order_by(order_by)
+
```
Test

danielfmiranda commented on network-api/networkapi/templates/pages/buyersguide/product_page.html:
```diff
@@ -168,21 +168,14 @@ <h2 class="tw-h3-heading">{% trans "Tips to protect yourself" %}</h2>
-                    </section>
-                {% endif %}
-            {% endwith %}
+            {% block related_products %}
+                {% with related_products=product.localized_related_products %}
+                    {% if related_products %}
+                        {% include "fragments/buyersguide/related_products_section.html" with related_products=related_products %}
+                    {% endif %}
+                {% endwith %}
+            {% endblock related_products %}
+
```
Updated the original product page template to take advantage of the new `related_products_section` template

danielfmiranda commented:
Thanks for the feedback @jhonatan-lopes! I have updated this PR accordingly and have also added some tests. This PR is now ready for re-review whenever you have a free moment 👍

jhonatan-lopes commented on network-api/networkapi/utility/orderables.py:
```diff
@@ -1,15 +1,21 @@
 
-def get_related_items(queryset: "models.QuerySet", related_item_field: str) -> list["models.Model"]:
+def get_related_items(
+    queryset: "models.QuerySet", related_item_field: str, order_by: Optional[str] = None
```
We're on Python 3.11 now, I think it's better if we use `str | None` for typing of optional paramaters

danielfmiranda replied:
Updated!

jhonatan-lopes commented on network-api/networkapi/wagtailpages/pagemodels/buyersguide/products.py:
```diff
@@ -1027,6 +1030,18 @@ def localized_related_products(self):
+        )
+        # This method, while inneficient, is neccesary as we are using it to render
+        # any related_product updates that are stored in memory, but not yet saved in the DB.
+        return [product.localized for product in related_products]
```
Now that I'm thinking about it, the "related_products" selection only occurs in the English/default version. The other versions are synced to the default location, so they don't have a "preview" window. I think it's safe to drop the `.localized` here

jhonatan-lopes commented on network-api/networkapi/wagtailpages/tests/buyersguide/test_products.py:
```diff
@@ -121,6 +121,189 @@ def test_localized_related_products_non_default_locale(self):
+
+        self.assertEqual(len(related_preview_products), 3)
+        # Order should be product 2, product 3, product 1.
+        self.assertListEqual(
```
Would it be better to use `assertCountEqual`?

danielfmiranda replied:
Hey Jhonatan! I actually removed asserts relating to count in my latest commit. I will resolve this conversation for now, but can reopen/add more updates if we choose to re-implement them.

jhonatan-lopes commented on network-api/networkapi/wagtailpages/tests/buyersguide/test_products.py:
```diff
@@ -121,6 +121,189 @@ def test_localized_related_products_non_default_locale(self):
+            [related_product_1, related_product_2, related_product_3],
+        )
+
+    def test_preview_related_products_with_non_default_locale(self):
```
We might not need this anymore, see above

jhonatan-lopes commented on network-api/networkapi/wagtailpages/tests/buyersguide/test_products.py:
The tests are great but one thing I'm worried is that they're all still hitting the database. I'm thinking that we need to test the `ProductPage.serve_preview` method perharps (https://github.com/wagtail/wagtail/blob/502dd7c723a3e0740915803b8fab6b9814bb9a9d/wagtail/models/__init__.py#L796-L812).

We can then manipulate the models using the [`modelcluster`](https://github.com/wagtail/django-modelcluster) syntax directly and without saving in the database and re-check the template response to make sure that the related products section order has updated properly.

jhonatan-lopes reviewed (changes requested):
Hey @danielfmiranda, thanks for this work! It's looking great, we just need to iron out some details

danielfmiranda commented:
Hi @jhonatan-lopes! Thanks for taking a look at this PR and for the great feedback.

I was unaware that pages had a `serve_preview` method. 💡 

I ended up using the [`page.make_preview_request`](https://github.com/wagtail/wagtail/blob/502dd7c723a3e0740915803b8fab6b9814bb9a9d/wagtail/models/__init__.py#L644) method, as it does most of the request-generation work that we would need to do in order to use `serve_preview`, and then also invokes the `serve_preview` command itself.

I have also updated my tests to use `modelcluster` syntax to the best of my ability, however I will admit I had some trouble, as it looks like we use an extra "through" model for this relationship.

Can you please take a look and let me know what you think? 

Thanks again!

danielfmiranda commented on network-api/networkapi/templates/fragments/buyersguide/related_product.html:
```diff
@@ -13,7 +13,7 @@
         </div>
         <p class="tw-body-small mt-3 mb-1">{{ related_product.company }}</p>
-        <p>{{ related_product.title }}</p>
+        <p class="product-title">{{ related_product.title }}</p>
```
Updated for use in tests

danielfmiranda commented on network-api/networkapi/utility/orderables.py:
```diff
@@ -4,12 +4,18 @@
 
-def get_related_items(queryset: "models.QuerySet", related_item_field: str) -> list["models.Model"]:
+def get_related_items(
+    queryset: "models.QuerySet", related_item_field: str, order_by: str | None = None
```
Updated to use str | None

jhonatan-lopes reviewed (approved):
LGTM :rocket: thanks @danielfmiranda
